src/cardano/CardanoEngine.ts

```diff
@@ -0,0 +1,556 @@
+      nativeAmount, // nativeAmount
+      networkFee: nativeNetworkFee, // networkFee, supposedly fixed
+      otherParams, // otherParams
+      ourReceiveAddresses: [], // ourReceiveAddresses
+      signedTx: '', // signedTx
+      txid: '', // txid
+      walletId: this.walletId
+    }
+
+    return edgeTransaction
+  }
+
+  async signTx(
+    edgeTransaction: EdgeTransaction,
+    privateKeys: JsonObject
+  ): Promise<EdgeTransaction> {
+    const { unsignedTx } = asCardanoTxOtherParams(edgeTransaction.otherParams)
+
+    const { mnemonic } = asCardanoPrivateKeys(this.currencyInfo.pluginId)(
+      privateKeys
+    )
+    const { accountKey } = this.tools.derivePrivateKeys(mnemonic)
+    const paymentKey = accountKey.derive(0).derive(0)
+
+    const txBody = Cardano.TransactionBody.from_hex(unsignedTx)
+    const txHash = Cardano.hash_transaction(txBody)
+    const witnesses = Cardano.TransactionWitnessSet.new()
+    const vkeyWitnesses = Cardano.Vkeywitnesses.new()
+    const vkeyWitness = Cardano.make_vkey_witness(
+      txHash,
+      paymentKey.to_raw_key()
+    )
+    vkeyWitnesses.add(vkeyWitness)
+    witnesses.set_vkeys(vkeyWitnesses)
+
+    const transaction = Cardano.Transaction.new(
+      txBody,
+      witnesses,
+      undefined // transaction metadata
+    )
+
+    edgeTransaction.signedTx = transaction.to_hex()
+    return edgeTransaction
+  }
+
+  async broadcastTx(
+    edgeTransaction: EdgeTransaction
+  ): Promise<EdgeTransaction> {
+    if (edgeTransaction.signedTx == null) throw new Error('Missing signedTx')
+
+    try {
+      const raw = await this.fetchPostAuthenticated(
+        'submittx',
+        base16.parse(edgeTransaction.signedTx)
+      )
+      const txid = asMaybe(asString)(raw)
+      if (txid == null) throw new Error(String(raw))
+      edgeTransaction.txid = txid
+      edgeTransaction.date = Date.now() / 1000
+      this.warn(`SUCCESS broadcastTx\n${cleanTxLogs(edgeTransaction)}`)
+    } catch (e: any) {
+      this.warn('FAILURE broadcastTx failed: ', e)
+      throw e
+    }
+
+    return edgeTransaction
+  }
+
+  async getFreshAddress(_options: any): Promise<EdgeFreshAddress> {
+    const { bech32Address } = asSafeCardanoWalletInfo(this.walletInfo).keys
+
+    return {
+      publicAddress: bech32Address
+    }
+  }
+}
+
+export async function makeCurrencyEngine(
+  env: PluginEnvironment<CardanoNetworkInfo>,
+  tools: CardanoTools,
+  walletInfo: EdgeWalletInfo,
+  opts: EdgeCurrencyEngineOptions
+): Promise<EdgeCurrencyEngine> {
+  const { initOptions } = env
+
+  const safeWalletInfo = asSafeCardanoWalletInfo(walletInfo)
+  const engine = new CardanoEngine(
+    env,
+    tools,
+    safeWalletInfo,
+    initOptions,
+    opts
+  )
+
+  // Do any async initialization necessary for the engine
+  await engine.loadEngine()
+
+  return engine
+}
+
+export const processCardanoTransaction = (opts: {
+  currencyCode: string
+  publicKey: string
+  tx: ReturnType<typeof asKoiosTransactionsRes>[number]
```

Comment:
Should you define a `type KoiosNetworkTx = ReturnType<typeof asKoiosTransactionsRes>[number]` in the types file?